File: NeetCode 150/Backtracking/78. Subsets.py
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        #Solution model 2: loop. Used in other problems as well
        res = []
        subset = []

        def backtrack(start):
            # Every path is a valid subset
            res.append(subset[:])

            for i in range(start,len(nums)):
                subset.append(nums[i])
                backtrack(i+1)
                subset.pop()
        backtrack(0)
        return res
        #Solution 1
        res = []
        subset = []

        def backtrack(i):
            if i == len(nums):
                res.append(subset.copy())
                return
            
            #include nums[i]
            subset.append(nums[i])
            backtrack(i+1)
            
            #exclude nums[i]
            subset.pop()
            backtrack(i+1)

        backtrack(0)
        return res
        

File: NeedCode 150 (JAVA)/Backtracking/78. Subsets.java
class Solution {
    private List<List<Integer>> res = new ArrayList<>();
    private List<Integer> subset = new ArrayList<>();

    public List<List<Integer>> subsets(int[] nums) {
        backtrack(0,nums);
        return res;
    }

    private void backtrack(int start, int[] nums){
        
        //every path is a valid path
        res.add(new ArrayList<>(subset)); // this will store a copy of subset like subset[:]

        for(int i = start;i<nums.length;i++){
            subset.add(nums[i]);
            backtrack(i+1,nums);
            subset.remove(subset.size() - 1);//this uses index and we are removing last element
        }
    }
}
